### PR TITLE
Snake game

### PR DESCRIPTION
**Pull request by paezao**

### CODE REVIEW

File: src/snake/main.rs [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+struct Food {
+    x: u16,
+    y: u16,
+}
+
+impl BodyPart {
+    fn crawl(&self) -> BodyPart {
+        let mut x = self.x;
+        let mut y = self.y;
+        let new_direction: Direction;
```
Comment (ticki) [inline src/snake/main.rs]: You don't need to annotate the type.

File: src/snake/main.rs [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+        match self.direction {
+            Direction::Up => { 
+                y += 1;
+                new_direction = Direction::Up;
```
Comment (ticki) [inline src/snake/main.rs]: Why this? Why not just directly use `self.direction` when returning?

File: src/snake/main.rs [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    /// Snake
+    snake: Snake,
+    /// Snake's Food
+    food: Food,
+    /// Speed
+    speed: u64,
+    /// Game Score
+    score: i32,
+    /// Interval between frames
+    interval: u64,
```
Comment (ticki) [inline src/snake/main.rs]: Is it really necessary to store this in the struct? Also, why not use `Duration`, instead?

File: src/snake/main.rs [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+}
+
+/// Snake's Food
+struct Food {
+    x: u16,
+    y: u16,
+}
+
+impl BodyPart {
+    fn crawl(&self) -> BodyPart {
```
Comment (ticki) [inline src/snake/main.rs:34]: Could you add docs to this one?

File: src/snake/main.rs [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+
+        let mut before = Instant::now();
+
+        loop {
+            self.interval = 1000 / self.speed;
+            let now = Instant::now();
+            let dt = (now.duration_since(before).subsec_nanos() / 1_000_000) as u64;
+
+            if dt < self.interval {
+                std::thread::sleep(Duration::from_millis(self.interval - dt));
```
Comment (ticki) [inline src/snake/main.rs]: You should import the `thread` module.

File: src/snake/main.rs [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+                continue;
+            }
+
+            before = now;
+
+            if !self.update() {
+                return;
+            }
+
+            if self.check_game_over() {
```
Comment (ticki) [inline src/snake/main.rs:118]: Fold this to the next line, through `&&`.

File: src/snake/main.rs [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+
+        self.write_rand(key_bytes[0]);
+
+        match key_bytes[0] {
+            b'q' => return false,
+            b'k' => self.turn_snake(Direction::Up),
+            b'j' => self.turn_snake(Direction::Down),
+            b'h' => self.turn_snake(Direction::Left),
+            b'l' => self.turn_snake(Direction::Right),
+            b'0' => {},
```
Comment (ticki) [inline src/snake/main.rs]: This is redundant.

File: src/snake/main.rs [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+            (_, 0) => true,
+            (x, _) if x == self.width as u16 => true,
+            (_, y) if y == self.height as u16 - 1 => true,
+            _ => false,
+        }
+    }
+
+    /// Grows the Snake's tail
+    fn grow_snake(&mut self) {
+        let (x, y, direction) = {
```
Comment (ticki) [inline src/snake/main.rs]: Why this? Why not instead declare the variable in a normal manner?

File: src/snake/main.rs [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+        });
+    }
+
+    /// Checks if the Snake is overlapping the food
+    fn check_eating(&mut self) -> bool {
+        let head = &self.snake.body.back().unwrap();
+        if (head.x, head.y) == (self.food.x, self.food.y) {
+            return true;
+        }
+        false
```
Comment (ticki) [inline src/snake/main.rs]: Fold this in an `if-else`.

File: src/snake/main.rs [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+    }
+
+    fn clear_snake(&mut self) {
+        for part in &self.snake.body {
+            self.stdout.goto(part.x, part.y).unwrap();
+            self.stdout.write(" ".as_bytes()).unwrap();
```
Comment (ticki) [inline src/snake/main.rs]: Use `b" "` instead.

File: src/snake/main.rs [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+        }
+    }
+
+    fn move_snake(&mut self) {
+        {
+            let tail = self.snake.body.pop_front().unwrap();
+            self.stdout.goto(tail.x, tail.y).unwrap();
+            self.stdout.write(" ".as_bytes()).unwrap();
```
Comment (ticki) [inline src/snake/main.rs]: Use `b" "` instead.

File: src/snake/main.rs [hunk 12]
```diff
@@ -0,0 +1,470 @@
...
+            
+            match self.snake.direction {
+                Direction::Up => (head.x, head.y - 1, Direction::Up),
+                Direction::Down => (head.x, head.y + 1, Direction::Down),
+                Direction::Left => (head.x - 1, head.y, Direction::Left),
+                Direction::Right => (head.x + 1, head.y, Direction::Right),
+            }
+        };
+
+        self.snake.body.push_back(BodyPart{
```
Comment (ticki) [inline src/snake/main.rs]: Formatting error here.

File: src/snake/main.rs [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+            x: x,
+            y: y,
+            direction: direction
+        });
+    }
+
+    fn turn_snake(&mut self, direction: Direction) {
+        match direction {
+            Direction::Up if self.snake.direction == Direction::Down => return,
```
Comment (ticki) [inline src/snake/main.rs]: You can fold these toghether to a single if clause.

File: src/snake/main.rs [hunk 14]
```diff
@@ -0,0 +1,470 @@
...
+        self.stdout.write_fmt(format_args!("SCORE: {}", self.score)).unwrap();
+        self.stdout.flush().unwrap();
+
+        loop {
+            // Repeatedly read a single byte.
+            let mut buf = [0];
+            self.stdin.read(&mut buf).unwrap();
+
+            match buf[0] {
+                b'r' => {
```
Comment (ticki) [inline src/snake/main.rs]: No need for brackets.

File: src/snake/main.rs [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+
+    /// Move the snake's food.
+    fn move_food(&mut self) {
+        loop {
+            let x = (self.read_rand() as u16 % (self.width as u16 - 2)) + 1;
+            let y = (self.read_rand() as u16 % (self.height as u16 - 2)) + 1;
+
+            if self.snake.body.iter().filter(|part| {
+                (x, y) == (part.x, part.y)
+            }).count() > 0 {
```
Comment (ticki) [inline src/snake/main.rs]: Just use `.next().is_some()`.

File: src/snake/main.rs [hunk 16]
```diff
@@ -0,0 +1,470 @@
...
+            }
+        };
+    }
+
+    /// Draws the snake's food.
+    fn draw_food(&mut self) {
+        self.stdout.reset().unwrap();
+
+        self.stdout.goto(self.food.x, self.food.y).unwrap();
+        self.stdout.write(b"*").unwrap();
```
Comment (ticki) [inline src/snake/main.rs]: Maybe use a more visible character?

File: src/snake/main.rs [hunk 17]
```diff
@@ -0,0 +1,470 @@
...
+
+        self.stdout.flush().unwrap();
+        self.stdout.reset().unwrap();
+    }
+
+    /// Draws the snake.
+    fn draw_snake(&mut self) {
+        self.stdout.reset().unwrap();
```
Comment (ticki) [inline src/snake/main.rs:357]: I am unsure why you reset both before and after the function?

File: src/snake/main.rs [hunk 18]
```diff
@@ -0,0 +1,470 @@
...
+    stdout.flush().unwrap();
+
+    let mut game = Game {
+        width: width,
+        height: height,
+        stdin: stdin,
+        stdout: stdout,
+        snake: Snake {
+            direction: Direction::Right,
+            body: vec![].into_iter().collect(),
```
Comment (ticki) [inline src/snake/main.rs]: Use `VecDeque::new()` instead.

**Comment by paezao** [comment]
Hey @Ticki, can you check out my changes?

**Comment by ticki** [comment]
Sure!

File: src/snake/main.rs [hunk 19]
```diff
@@ -0,0 +1,464 @@
...
+    /// Snake
+    snake: Snake,
+    /// Snake's Food
+    food: Food,
+    /// Speed
+    speed: u64,
+    /// Game Score
+    score: i32,
+    /// This will be modified when a random value is read or written.
+    seed: usize,
```
Comment (ticki) [inline src/snake/main.rs]: You could use the `Randomizer` struct from libextra, see minesweeper.

File: src/snake/main.rs (hunk 6 above)
Comment (ticki) [inline src/snake/main.rs:118]: This still applies :point_up_2:

File: src/snake/main.rs [hunk 20]
```diff
@@ -0,0 +1,464 @@
...
+            b'l' => self.turn_snake(Direction::Right),
+            _ => {},
+        }
+
+        self.move_snake();
+
+        true
+    }
+
+    /// Read a number from the randomizer.
```
Comment (ticki) [inline src/snake/main.rs]: Again, these can be replaced by what is provided by `libextra`.

File: src/snake/main.rs [hunk 21]
```diff
@@ -0,0 +1,464 @@
...
+    fn check_game_over(&mut self) -> bool {
+        let head = &self.snake.body.back().unwrap();
+
+        if self.snake.body.iter().filter(|part| {
+            (head.x, head.y) == (part.x, part.y)
+        }).count() > 1 {
+            return true;
+        }
+
+        match (head.x, head.y) {
```
Comment (ticki) [inline src/snake/main.rs]: No need for matching, you can do tuple indexing instead:

```
self.0 ==0 || self.1 == 0 | self.0 == self.width as u16 || self.1 == self.height as u16
```

File: src/snake/main.rs (hunk 8 above)
Comment (ticki) [inline src/snake/main.rs]: This still applies.

File: src/snake/main.rs (hunk 10 above)
Comment (ticki) [inline src/snake/main.rs]: :point_up_2:

**Comment by ticki** [comment]
Simply beautiful.

File: src/snake/main.rs [hunk 22]
```diff
@@ -221,12 +209,11 @@ impl<R: Read, W: Write> Game<R, W> {
...
         }
 
-        match (head.x, head.y) {
-            (0, _) => true,
-            (_, 0) => true,
-            (x, _) if x == self.width as u16 => true,
-            (_, y) if y == self.height as u16 - 1 => true,
-            _ => false,
+        if head.x == 0 || head.y == 0 || head.x == self.width as u16 || head.y == self.height as u16 - 1 {
+            true
```
Comment (ticki) [inline src/snake/main.rs]: You did the deadly sin of programming.

File: src/snake/main.rs [hunk 23]
```diff
@@ -0,0 +1,452 @@
...
+    }
+
+    /// Check if the Snake is overlapping a wall or a body part
+    fn check_game_over(&mut self) -> bool {
+        let head = &self.snake.body.back().unwrap();
+
+        if self.snake.body.iter().filter(|part| {
+            (head.x, head.y) == (part.x, part.y)
+        }).count() > 1 {
+            return true;
```
Comment (ticki) [inline src/snake/main.rs]: Don't do this, just return the condition itself.

File: src/snake/main.rs [hunk 24]
```diff
@@ -0,0 +1,452 @@
...
+            x: x,
+            y: y,
+            direction: direction,
+        });
+    }
+
+    /// Checks if the Snake is overlapping the food
+    fn check_eating(&mut self) -> bool {
+        let head = &self.snake.body.back().unwrap();
+        if (head.x, head.y) == (self.food.x, self.food.y) {
```
Comment (ticki) [inline src/snake/main.rs]: And deadly sins here too.

**Comment by ticki** [comment]
Looks awesome! Will merge when Travis is done building.

**Pull request merged**